### ISSUE TITLE
Ingester stalled and OOMed after compaction

### ISSUE THREAD

**Issue by cchen1103** [grafana/mimir#3456]
### What did you do?

We are running Mimir 2.3.1 in AWS EKS, the ingester is set using EBS io1 with 2GB/iops. We notice that after compaction every 2 hours, some of the ingesters has a very low throughput. Most of query/push requests finished over 1 minute, and only small portion of requests are completed within 20ms.

The memory usage is slow build up along with the build up on inflight requests. We believe the memory usage buildup is due to the inflight request building up. The extremely process on the requests (suspect the push requests) causes the inflight request build up.

The CPU utilization keep high after compaction. After 1 hour 30 minutes, the memory usage hit the system limit and get restarted.

After the restart of ingester, it seems to be normalized. The cycle will begin after compaction.

### What did you expect to see?

After compaction, ingester performs push/query request with low latency as it used to be.

### What did you see instead? Under which circumstances?

We notice that after compaction every 2 hours, some of the ingesters has a very low throughput. Most of query/push requests finished over 1 minute, and only small portion of requests are completed within 20ms.

### System information

Ubunto 20.4-lts

### Prometheus version

```text
Mimir: 2.3.1
```

**Comment by pracucci** [comment]
👋 I don't have enough information to understand what's going on. I'm wondering if this is caused by exausting the disk write buffer. Few initial questions / suggestions:
1. How many in-memory series do you have per ingester right before compation? How much memory available to the pod?
2. is the happing at compaction (UTC hour is odd, so `01:00`, `03:00`, ...) or at even hours UTC (so `00:00`, `01:00`)?
3. What's actual disk I/O and latency when this issue happens?

**Comment by cchen1103** [comment]
1. We have 2-3 Mil of series per ingester before compaction. There are 64G memory on the server and it generally uses 32-34 G memory.
2. The compaction is happening is odd number, such as 1:00, 3:00 ... the time is UTC.
3. The disk write rate is about 15M/s per pod on ingester, the read is small since the compaction does not involve much read. The ingester disk utilization is about 15%.

Thanks for looking into this issue.
<img width="810" alt="Screenshot 2022-11-16 at 10 08 04 AM" src="https://user-images.githubusercontent.com/10091716/202260588-f7173c74-7267-4777-bebe-111792fa5c10.png">
<img width="810" alt="Screenshot 2022-11-16 at 10 08 12 AM" src="https://user-images.githubusercontent.com/10091716/202260593-d391f7dc-39f7-463f-b79f-450f92967180.png">
<img width="1624" alt="Screenshot 2022-11-16 at 10 13 21 AM" src="https://user-images.githubusercontent.com/10091716/202260597-e01d7a3c-49f8-44ea-92c7-65e330ec4732.png">
<img width="807" alt="Screenshot 2022-11-16 at 10 07 53 AM" src="https://user-images.githubusercontent.com/10091716/202260634-946c7ce4-a4c5-4736-a398-b2014083ac8d.png">

**Comment by cchen1103** [comment]
I add the following metrics on ingester when the issue happened. the compaction started at 9:00 AM, and after the compaction, some of the ingesters get OOMed due to build up in-flight requests. The total ingestion volume or traffic pattern are not changed before and after compaction.
<img width="1648" alt="Screenshot 2022-11-16 at 2 08 45 PM" src="https://user-images.githubusercontent.com/10091716/202305567-481d71ca-3850-4ded-bae6-661e7e7b6565.png">
<img width="1623" alt="Screenshot 2022-11-16 at 2 09 05 PM" src
<img width="1646" alt="Screenshot 2022-11-16 at 2 07 16 PM" src="https://user-images.githubusercontent.com/10091716/202305589-0859e5ca-a6c0-4740-b008-5a0d61803aae.png">
<img width="1633" alt="Screenshot 2022-11-16 at 2 10 10 PM" src="https://user-images.githubusercontent.com/10091716/202305591-a803c192-20db-4870-8d7e-3f6fb4e5a45f.png">
="https://user-images.githubusercontent.com/10091716/202305577-cc95355e-7672-419b-8213-0efa038fa23e.png">
<img width="1643" alt="Screenshot 2022-11-16 at 2 09 15 PM" src="https://user-images.githubusercontent.com/10091716/202305580-8ce38157-728d-4700-9cac-df396db39a69.png">
<img width="1631" alt="Screenshot 2022-11-16 at 2 09 25 PM" src="https://user-images.githubusercontent.com/10091716/202305582-6b0d1e70-b8de-419b-a3d5-0e4050db1b1c.png">

**Comment by pracucci** [comment]
Thanks for sharing all such details. Looking at these metrics I agree with your conclusion that ingesters went OOM because of the higher inflight requests, but it's not super clear to me why you've experienced such an high number of inflight requests for an extended period of time.

The TSDB head compaction is expected to take at most few minutes (which is also what I can see from disk I/O graphs) but the high memory and high in-flight requests last way longer, and in some ingesters don't even start at the TSDB head compaction time but later.

**Comment by cchen1103** [comment]
I agree with your assessment. I spent quite some time looking into this as well. The throughput after compaction suddenly get into very low. I do not think it is related to capacity or disk i/o constrains. Instead, it seems to be either program get into thread dead lock or significantly reduces go routine available for the application.

The compaction/shipper and creating chunk head uses mmap. I wonder if the mmap (which is synchronized) cause the lock up or major page failure cause the slowness. Keep in mind that the query is running against ingester all the time as well.

**Comment by cchen1103** [comment]
From our log analysis, this piece of code https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L646-L779 is where it get very very slow response. We saw normal call comes in and error out within 10 ns at line #646, but we do not see the regular error log volumes within milliseconds response reaching at line #779. I am pretty confidently say that the problem throttled part lies before these part of code.

**Comment by Logiraptor** [comment]
This may be unrelated, but this sounds a lot like an incident I responded to. In that case, we had many ingesters packed onto the same underlying k8s nodes. Each ingester was requesting some CPU (say 5 for the example), and using those 5 cores during normal operation. Once head compaction started, all ingesters immediately needed 7-8 cores for a short period, but because we had many ingesters packed onto the same node, they were being throttled and fighting for CPU. This dramatically increased the ingester latency and triggered write failures. In our case we just needed to increase the CPU request for ingesters to make sure we had extra CPU headroom during compaction. This also naturally spread ingesters across more nodes.

Can you check your ingester CPU usage vs requests and limits before / during compaction? Also check the underlying node CPU before / during compaction.

**Comment by cchen1103** [comment]
The ingester CPU are at 15 vCPU. All ingesters are allocated on different instances. There is no significant CPU utilization increase according to the metrics. 

One thing that I notice that this compaction issue happens when we have consistent number of query running (from rulers), and some of the query is quite heavy (longer response time)
